### PR TITLE
[csp] Wait for all tests to be defined

### PR DESCRIPTION
**Pull request by jugglinmike**
Configure the test harness to delay completion until all
asynchronously-defined tests are available.

### CODE REVIEW

File: content-security-policy/generic/generic-0_2.html
Comment (andypaicu) [inline content-security-policy/generic/generic-0_2.html]: This expectation seems to have flipped, was this test incorrect to begin with?

File: content-security-policy/reporting/report-uri-from-child-frame.html
Comment (andypaicu) [inline content-security-policy/reporting/report-uri-from-child-frame.html]: I believe this was here precisely to have a declared test so the testing harness won't immediately conclude the test when the page loads. I think this test now will succeed immediately without waiting for a message.

File: content-security-policy/reporting/report-only-in-meta.sub.html
Comment (andypaicu) [inline content-security-policy/reporting/report-only-in-meta.sub.html]: Is there a chance of the script element onload event being fired before we set the handler? Could we move the document.body.appendChild(script); line after this?

Comment (jugglinmike) [reply content-security-policy/reporting/report-only-in-meta.sub.html]: DOM event handlers do not run in parallel, but I'm happy to relocate if you feel it improves readability.

File: content-security-policy/reporting/report-uri-from-child-frame.html
Comment (jugglinmike) [reply content-security-policy/reporting/report-uri-from-child-frame.html]: In the absence of any tests, testharness.js will not react to the `load` event. Instead, it will wait for the global `done` function to be invoked (as an indication of a single-page test) or for a sub-test to be defined. This test defines one sub-test via `checkReport.sub.js`, and since that is the only test it defines, there's no need to explicitly manage scheduling.

More than being superfluous, the removed sub-test was a source of instability. It completed before the meaningful sub-test was defined, so there was a race condition regarding when testharness.js determined completion.

File: content-security-policy/generic/generic-0_2.html
Comment (jugglinmike) [reply content-security-policy/generic/generic-0_2.html]: Yes, it was incorrect. There are no CSP violations, so the event should not have been emitted at all (regardless of the value of the `violatedDirective` property). Also, the sub-test was written to complete in two different ways: via a subsequent inline script and via the `step_func_done` method.

File: content-security-policy/reporting/report-uri-from-child-frame.html
Comment (andypaicu) [reply content-security-policy/reporting/report-uri-from-child-frame.html]: Ah ok, I did not know it behaves like that when there are no tests.

File: content-security-policy/reporting/report-only-in-meta.sub.html
Comment (andypaicu) [reply content-security-policy/reporting/report-only-in-meta.sub.html]: I think it's nice to set all the properties on the element before adding it to the DOM but it does not really matter.

**Review by andypaicu — APPROVED** [review]

File: content-security-policy/generic/generic-0_2.html
Comment (andypaicu) [reply content-security-policy/generic/generic-0_2.html]: Ok

**Comment by jugglinmike** [comment]
Thanks for the review, @andypaicu! Would you mind merging this?

**Comment by jugglinmike** [comment]
@andypaicu seems like this may have slipped off your radar. Would you mind merging it?

**Comment by foolip** [comment]
@jugglinmike, if approved please do merge yourself, it reduces round-trips and would have avoided the possibility of me trying to fix the same bug.

**Comment by jugglinmike** [comment]
@foolip I interpret an approval that isn't immediately followed by merging to mean that further review is expected. If there is substantive feedback or merge conflicts, it's more clear that the ball's in my court. That's not the case here, though. This isn't the first time I've seen this in WPT, so I may be misunderstanding something. How do you interpret this situation?

**Comment by foolip** [comment]
@jugglinmike this is a matter of convention and not written down anywhere, but it's very common for people to approve with no comment without merging. I see @Ms2ger do it, and I do it. IMHO it's a good idea to do so if you're not sure if the author might have hoped for someone else to review. Approving with nits and trusting the author to fix those and then merge is also common, typically as "LGTM % nits".

Basically if merging is technically possible it's OK. Reviewers will learn to not approve if that's not what they mean :)

**Comment by jugglinmike** [comment]
Got it. I'll miss having the approval memorialized in the git history, but I'm probably overthinking things. Thanks for the explanation!

**Comment by foolip** [comment]
I love a good git history, but for this project I don't have high hopes... We have loads of automatic exports with commit messages that are nonsensical taken out of the original context. And people squash-merge PRs without cleaning up the commit messages, making them just a bullet list of somewhat random phrases. Often the useful explanation is in PRs and not in the commit messages.

The wrong committer won't stand out much against this background :D

**Pull request merged**
